### ISSUE TITLE
ECONNREFUSED when docker instance queries docker neo4j

### ISSUE THREAD

**Issue by AdrienLemaire** [neo4j/docker-neo4j#267]
**Expected behavior**: My API server should be able to do unencrypted requests to the neo4j docker instance at `bolt://neo4j:57687` (mapped to `bolt://172.20.0.2:57687`)

**Actual behavior**:
```
api_1    | Failed to connect to server. Please ensure that your database is listening on the correct host and port and that you have compatible encryption settings both on Neo4j server and driver. Note that the default encryption setting has changed in Neo4j 4.0. Caused by: connect ECONNREFUSED 172.20.0.2:57687
```

**Additional info**:

OS: Arch linux
Running neo4j:4.1-enterprise
```dockerfile
# https://hub.docker.com/_/neo4j/
FROM neo4j:4.1-enterprise

# https://neo4j.com/docs/labs/apoc/current/introduction/#installation
# https://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/
ENV APOC_VERSION=4.1.0.0

ENV APOC_URI=https://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/download/${APOC_VERSION}/apoc-${APOC_VERSION}-all.jar
##curl is removed during the build of neo4j v4.0, I guess because of its size and dependencies
#RUN sh -c 'cd /var/lib/neo4j/plugins && curl -L -O "${APOC_URI}"'
RUN sh -c 'wget -P /var/lib/neo4j/plugins/ "${APOC_URI}"'


EXPOSE 7474 7473 7687

CMD ["neo4j"]
```

Docker:
```console
docker version
Client:
 Version:           19.03.12-ce
 API version:       1.40
 Go version:        go1.14.5
 Git commit:        48a66213fe
 Built:             Sat Jul 18 01:33:21 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.12-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.14.5
  Git commit:       48a66213fe
  Built:            Sat Jul 18 01:32:59 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.3.4.m
  GitCommit:        d76c121f76a5fc8a462dc64594aea72fe18e1178.m
 runc:
  Version:          1.0.0-rc91
  GitCommit:        24a3cf88a7ae5f4995f6750654c0e2ca61ef4bb2
 docker-init:
  Version:          0.18.0
```

docker-compose.yml
```yml
---
version: '3.8'

services:

  neo4j:
    build: ./neo4j_db
    ports:
      - 57473:7473  # HTTPS
      - 57474:7474  # HTTP
      - 57687:7687  # BOLT
    environment:
      #  you can override every neo4j.conf property by prefix it with NEO4J_
      #  https://neo4j.com/docs/labs/neo4j-streams/current/docker/#neo4j_streams_docker
      - NEO4J_dbms_default__listen__address=0.0.0.0
      - NEO4J_dbms_default__advertised__address=neo4j
      - NEO4J_dbms_connector_bolt_advertised__address=:57687
      - NEO4J_dbms_ssl_policy_bolt_client__auth=NONE
      - NEO4J_dbms_ssl_policy_https_client__auth=NONE
      - NEO4J_dbms_security_procedures_unrestricted=apoc.*
      - NEO4J_apoc_import_file_enabled=true
      - NEO4J_apoc_export_file_enabled=true
      - NEO4J_ACCEPT_LICENSE_AGREEMENT=yes
      - NEO4J_AUTH=none

  api:
    build: ./api_server
    ports:
      - 54001:4001
    environment:
      - NEO4J_URI=bolt://neo4j:57687  # https://neo4j.com/docs/driver-manual/current/client-applications/#driver-configuration-examples
      - NEO4J_USER=neo4j
      - GRAPHQL_LISTEN_PORT=4001
      - GRAPHQL_URI=http://api:54001/graphql
    links:
      - neo4j
    depends_on:
      - neo4j

networks:
  default:
```

I have no problem connecting to the neo4j docker db from my host by querying localhost
```console
$ cypher-shell -a 127.0.0.1:57687 -u neo4j -panythingWillWork
Connected to Neo4j 4.1.1 at bolt://127.0.0.1:57687 as user neo4j.
Type :help for a list of available commands or :exit to exit the shell.
Note that Cypher queries must end with a semicolon.
neo4j@neo4j> 
```
I could also successfully run an API server from my host, and connect to the db using again localhost.
From the doc
> If routing traffic via a proxy, or if port mappings are in use, it is possible to specify advertised_address for each connector individually. For example, if port 7687 on the Neo4j Server is mapped from port 9000 on the external network, > specify the advertised_address for the Bolt connector:
> dbms.connector.bolt.advertised_address=<server-name>:9000
I made sure that neo4j listens to 0.0.0.0 and has an advertised port of 57687

```console
$ docker-compose exec neo4j cat conf/neo4j.conf |  rg -v '^#' | rg '_address|bolt'
dbms.connector.bolt.enabled=true
dbms.ssl.policy.bolt.client_auth=NONE
dbms.default_listen_address=0.0.0.0
dbms.default_advertised_address=neo4j
dbms.connector.bolt.advertised_address=:57687
causal_clustering.transaction_advertised_address=0bf0ada64463:6000
causal_clustering.raft_advertised_address=0bf0ada64463:7000
causal_clustering.discovery_advertised_address=0bf0ada64463:5000
```
This is a simple neo4j db without causal clustering
![](https://i.imgur.com/T8kTWX8.png)

Any idea how to resolve this issue?

**Comment by jennyowen** [comment]
Hi, sorry for the delayed reply.

Since `cypher-shell` and the neo4j browser can connect to your neo4j container, that suggests to me that the problem might be with how your API is using the Neo4j drivers rather than with the docker image.

However, I'm confused from your description. 
1. Is the neo4j database running on `localhost` and your server is on a different machine and it can't connect? If so, are you able to connect from a remote machine manually using the browser or `cypher-shell`?
2. Are neo4j and your api server two separate containers running on the same machine? Your docker-compose suggests maybe? If you add both containers on to the `host` network does it work?

**Comment by AdrienLemaire** [comment]
@jennyowen thanks for the reply. I was busy with another project, so your timing is great :+1: 

Here's my setup:
* A system-wide installed Neo4j db running on `bolt://localhost:7687`
* A docker-compose environment containing a neo4j container and an API container, using the default network with port mappings such as 57687:7687.

From my system/host, when querying localhost:7687, I connect to my host neo4j db. When querying localhost:57687, I connect to my docker neo4j db.
The problem occurs when trying to connect to the docker neo4j from the docker API server container. Within a docker container, localhost resolves to that container's localhost network, which differs from my host localhost. docker-compose offers network aliases to query other containers (here, my API server can query the neo4j container through its `neo4j` network alias, which resolves to the IP 172.20.0.2). 
It appears as if the neo4j container refuses connections from the API server container because the referer IP is 172.20.0.2 instead of localhost...

**Comment by AdrienLemaire** [comment]
@jennyowen sorry, this was a stupid mistake :) Since I'm connecting to the neo4j container IP directly, there is no port mapping done. I only needed to connect to `bolt://neo4j:7687` instead of `bolt://neo4j:57687`. Problem resolved.
